### ISSUE TITLE
Xeon L3426 - Turbo boost not working

### ISSUE THREAD

**Issue by benwwchen** [Piker-Alpha/ssdtPRGen.sh#215]
Thanks for your hard work.

I'm using the script for my Xeon L3426. It should be able to turbo boost to 3.2GHz (1 core), 3.06GHz (2 cores) and 2.13GHz (3 or 4 cores) according to [this](http://www.cpu-world.com/CPUs/Xeon/Intel-Xeon%20L3426%20-%20BV80605004737AA%20%28BX80605L3426%29.html). And it can turbo boost correctly on Windows 10.

But before using this script, I can only get 2.13GHz (4 cores together) on El Capitan(P-State from 9 to 16):
(I disabled logMSRs in AppleIntelInfo.kext because it caused kernel panic maybe because my motherboard is an ASUS one.)

```
AppleIntelInfo.kext v1.2 Copyright © 2012-2015 Pike R. Alpha. All rights reserved

Settings:
------------------------------------
logMSRs............................: 0
logIGPU............................: 0
logIntelRegs.......................: 1
logCStates.........................: 1
logIPGStyle........................: 1
InitialTSC.........................: 0x506a41f3cb
MWAIT C-States.....................: 4384

CPU Ratio Info:
------------------------------------
CPU Low Frequency Mode.............: 900 MHz
CPU Maximum non-Turbo Frequency....: 1400 MHz
CPU Maximum Turbo Frequency........: 2400 MHz
CPU P-States [ 14 ]
CPU P-States [ 12 14 ]
CPU P-States [ 10 12 14 ]
CPU P-States [ 10 11 12 14 ]
CPU P-States [ 10 11 12 13 14 ]
CPU P-States [ 10 11 12 13 14 15 ]
CPU P-States [ 10 11 12 13 14 15 16 ]
CPU P-States [ 9 10 11 12 13 14 15 16 ]
```

Then I added

```
gUserDefinedCPUList=(
L3426,45,1200,1800,3200,4,8,2,133
)
```

in User Defined.cfg. Tried generating ssdt with SMBIOS MacPro5,1, Mac mini 5,1, iMac 11,1, iMac 12,1. All of them can't turbo boost at all:

```
AppleIntelInfo.kext v1.2 Copyright © 2012-2015 Pike R. Alpha. All rights reserved

Settings:
------------------------------------
logMSRs............................: 0
logIGPU............................: 0
logIntelRegs.......................: 1
logCStates.........................: 1
logIPGStyle........................: 1
InitialTSC.........................: 0x4ceff5263e
MWAIT C-States.....................: 4384

CPU Ratio Info:
------------------------------------
CPU Low Frequency Mode.............: 900 MHz
CPU Maximum non-Turbo Frequency....: 1400 MHz
CPU Maximum Turbo Frequency........: 2400 MHz
CPU P-States [ 12 ]
CPU C3-Cores [ 0 2 3 5 7 ]
CPU C6-Cores [ 0 1 2 3 4 5 6 7 ]
CPU C3-Cores [ 0 1 2 3 4 5 7 ]
CPU P-States [ 12 14 ]
CPU P-States [ 11 12 14 ]
CPU C3-Cores [ 0 1 2 3 4 5 6 7 ]
CPU P-States [ 10 11 12 14 ]
CPU P-States [ 9 10 11 12 14 ]
CPU P-States [ 9 10 11 12 13 14 ]
```

The ssdtPRGen.sh output is:
(I'm using v16.7, because beta gives compilation errors)

```
./ssdtPRGen.sh -p L3426
ssdtPRGen.sh v0.9 Copyright (c) 2011-2012 by † RevoGirl
             v6.6 Copyright (c) 2013 by † Jeroen
             v16.7 Copyright (c) 2013-2016 by Pike R. Alpha
-----------------------------------------------------------
Bugs > https://github.com/Piker-Alpha/ssdtPRGen.sh/issues <

Override value: (-p) processor model, now using: L3426!

System information: Mac OS X 10.11.4 (15E65)
Brandstring 'Intel(R) Xeon(R) CPU L3426 @ 1.87GHz'

Scope (_PR_) {590 bytes} with ACPI Processor declarations found in the DSDT (ACPI 1.0 compliant)
Generating ssdt.dsl for a 'Macmini5,1' with board-id [Mac-8ED6AF5B48C039E1]
Intel L3426 processor [0x106E5] setup [0x0501]
Processor matched!
With a maximum TDP of 45 Watt, as specified by Intel
Number logical CPU's: 8 (Core Frequency: 1800 MHz)
Number of Turbo States: 10 (1933-3200 MHz)
Number of P-States: 16 (1200-3200 MHz)
Injected C-States for P001 (C1,C3,C6)
Warning: 'cpu-type' may be set improperly (0x0501 instead of 0x0601)

Intel ACPI Component Architecture
ASL Optimizing Compiler version 20140926-64 [Nov  6 2014]
Copyright (c) 2000 - 2014 Intel Corporation

ASL Input:     /Users/benwwchen/Library/ssdtPRGen/ssdt.dsl - 202 lines, 5620 bytes, 48 keywords
AML Output:    /Users/benwwchen/Library/ssdtPRGen/ssdt.aml - 1425 bytes, 19 named objects, 29 executable opcodes

Compilation complete. 0 Errors, 0 Warnings, 0 Remarks, 0 Optimizations
```

By the way, Clover does recognize my CPU and all numbers seem correct:

```
2:899  0:000  CPU Vendor = 756E6547 Model=106E5
2:899  0:000  The CPU supported turbo
2:899  0:000  BrandString = Intel(R) Xeon(R) CPU           L3426  @ 1.87GHz
2:899  0:000  FSBFrequency=133MHz DMIvalue=133330kHz
2:899  0:000  Corrected FSBFrequency=133MHz
2:899  0:000  Vendor/Model/Stepping: 0x756E6547/0x1E/0x5
2:899  0:000  Family/ExtFamily: 0x6/0x0
2:899  0:000  MaxDiv/MinDiv: 14.0/9
2:899  0:000  Turbo: 16/16/23/24
2:899  0:000  Features: 0xBFEBFBFF
2:899  0:000  Threads: 8
2:899  0:000  Cores: 4
2:899  0:000  FSB: 133 MHz
2:899  0:000  CPU: 1866 MHz
2:899  0:000  TSC: 1866 MHz
2:899  0:000  PIS: 533 MHz
```

Here is the APIC, DSDT in ~/Library/ssdtPRGen and SSDT generated by ssdtPRGen.sh with MacPro5,1 SMBIOS:
[dsdt,apic,ssdt.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/233267/dsdt.apic.ssdt.zip)

Here is the SSDT generated by Clover which can turbo to 2.13 GHz:
[System SSDT.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/233309/System.SSDT.zip)

And this is the original ACPI tables extracted in Windows:
[ACPI.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/233312/ACPI.zip)

What can I do to get the turbo boost working? Thanks.

**Comment by benwwchen** [comment]
Here is the output of the latest beta version, which has compilation errors:

```
./ssdtPRGen.sh -p L3426


ssdtPRGen.sh v0.9  Copyright (c) 2011-2012 by † RevoGirl
             v6.6  Copyright (c) 2013 by † Jeroen
             v18.2 Copyright (c) 2013-2016 by Pike R. Alpha
-----------------------------------------------------------
Bugs > https://github.com/Piker-Alpha/ssdtPRGen.sh/issues <

System information: Mac OS X 10.11.4 (15E65)
Brandstring: "Intel(R) Xeon(R) CPU L3426 @ 1.87GHz"

Override value: (-p) processor model, now using: L3426!



Scope (_PR_) {592 bytes} with ACPI Processor declarations found in the DSDT (ACPI 1.0 compliant)
Generating ssdt.dsl for a 'iMac12,1' with board-id [Mac-942B5BF58194151B]
Intel L3426 processor [0x106E5] setup [0x0501]
With a maximum TDP of 45 Watt, as specified by Intel
Number logical CPU's: 8 (Core Frequency: 1867 MHz)
Number of Turbo States: 9 (2000-3192 MHz)
Number of P-States: 16 (1197-3192 MHz)
Injected C-States for P001 (C1,C3,C6)
Warning: 'cpu-type' may be set improperly (0x0501 instead of 0x0601)

Compiling: ssdt_pr.dsl
Intel ACPI Component Architecture
ASL Optimizing Compiler version 20140926-64 [Nov  6 2014]
Copyright (c) 2000 - 2014 Intel Corporation

Compiler aborting due to parser-detected syntax error(s)
/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     21:     External (\_PR_.001, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     21:     External (\_PR_.001, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     21:     External (\_PR_.001, DeviceObj)
Error    6126 -   Invalid character (0x01), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     22:     External (\_PR_.002, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     22:     External (\_PR_.002, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     22:     External (\_PR_.002, DeviceObj)
Error    6126 -   Invalid character (0x02), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     23:     External (\_PR_.003, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     23:     External (\_PR_.003, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     23:     External (\_PR_.003, DeviceObj)
Error    6126 -   Invalid character (0x03), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     24:     External (\_PR_.004, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     24:     External (\_PR_.004, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     24:     External (\_PR_.004, DeviceObj)
Error    6126 -   Invalid character (0x04), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     25:     External (\_PR_.005, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     25:     External (\_PR_.005, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     25:     External (\_PR_.005, DeviceObj)
Error    6126 -   Invalid character (0x05), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     26:     External (\_PR_.006, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     26:     External (\_PR_.006, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     26:     External (\_PR_.006, DeviceObj)
Error    6126 -   Invalid character (0x06), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     27:     External (\_PR_.007, DeviceObj)
Error    6126 -                                                       ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     27:     External (\_PR_.007, DeviceObj)
Error    6126 -  syntax error, unexpected PARSEOP_INTEGER, expecting ')' ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl     27:     External (\_PR_.007, DeviceObj)
Error    6126 -   Invalid character (0x07), expecting ASL keyword or name ^ 

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl    133:     Scope (\_PR_.001)
Error    6126 -                                                    ^ Invalid character (0x2E), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl    133:     Scope (\_PR_.001)
Error    6126 -                                                       ^ syntax error, unexpected PARSEOP_INTEGER, expecting ')'

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl    133:     Scope (\_PR_.001)
Error    6126 -                                                        ^ Invalid character (0x01), expecting ASL keyword or name

/Users/benwwchen/Library/ssdtPRGen/ssdt.dsl    143: 
Error    6126 - syntax error, unexpected PARSEOP_SCOPE, expecting $end and premature End-Of-File

ASL Input:     /Users/benwwchen/Library/ssdtPRGen/ssdt.dsl - 143 lines, 4594 bytes, 27 keywords

Compilation complete. 25 Errors, 0 Warnings, 0 Remarks, 0 Optimizations
Do you want to open ssdt.dsl (y/n)? n
```

And here is the ssdt.dsl generated: 
[ssdt.dsl.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/252334/ssdt.dsl.zip)

Then I fixed the syntax errors by changing the names of the CPUs, but still got no turbo states.
Here is the fixed ssdt.dsl: 
[ssdt_fixed.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/252602/ssdt_fixed.zip)

Is my CPU too old? Could you please help me with this or tell what I can do about it? Thanks.

**Comment by Piker-Alpha** [comment]
Sorry for being unresponsive, but our second child and my work for our new home are eating up all my free time. Anyway. I have good news. I figured out why the script failed. Thing is. Your DSDT includes Alias() statements (to add Processor labels) and that made it fail. The latest update (v18.3) now includes a new check and running this:

`./ssdtPRGen.sh -mode custom -p L3426`

Results is this (snippet) of the generated SSDT.dsl

Name (APLF, Zero)
Name (APSN, 0x0A)
Name (APSS, Package (0x10)
{
    Package (0x06) { 0x0C80, 0x00AFC8, 0x0A, 0x0A, 0x0018, 0x0018 },
    Package (0x06) { 0x0BFA, 0x00AFC8, 0x0A, 0x0A, 0x0017, 0x0017 },
    Package (0x06) { 0x0B75, 0x00AFC8, 0x0A, 0x0A, 0x0016, 0x0016 },
    Package (0x06) { 0x0AF0, 0x00AFC8, 0x0A, 0x0A, 0x0015, 0x0015 },
    Package (0x06) { 0x0A6A, 0x00AFC8, 0x0A, 0x0A, 0x0014, 0x0014 },
    Package (0x06) { 0x09E5, 0x00AFC8, 0x0A, 0x0A, 0x0013, 0x0013 },
    Package (0x06) { 0x0960, 0x00AFC8, 0x0A, 0x0A, 0x0012, 0x0012 },
    Package (0x06) { 0x08DB, 0x00AFC8, 0x0A, 0x0A, 0x0011, 0x0011 },
    Package (0x06) { 0x0855, 0x00AFC8, 0x0A, 0x0A, 0x0010, 0x0010 },
    Package (0x06) { 0x07D0, 0x00AFC8, 0x0A, 0x0A, 0x000F, 0x000F },
    Package (0x06) { 0x074B, 0x00AFC8, 0x0A, 0x0A, 0x000E, 0x000E },
    Package (0x06) { 0x06C5, 0x00AFC8, 0x0A, 0x0A, 0x000D, 0x000D },
    Package (0x06) { 0x0640, 0x008025, 0x0A, 0x0A, 0x000C, 0x000C },
    Package (0x06) { 0x05BB, 0x0065B9, 0x0A, 0x0A, 0x000B, 0x000B },
    Package (0x06) { 0x0536, 0x004F36, 0x0A, 0x0A, 0x000A, 0x000A },
    Package (0x06) { 0x04B0, 0x003415, 0x0A, 0x0A, 0x0009, 0x0009 }
})

The output of AppleIntelInfo.kext is wrong/was developed for Intel Core In processors with 100MHz banks and thus (for example) 9 should be read as: 9 x 133.3 MHz = 1199.7 MHz) as LFM (1200MHz). Now have a look at this output:

`CPU P-States [ 9 10 11 12 13 14 15 16 ]`

You should read the last two as:

15 x 133.3 = 2000 MHz
16 = 133.3 = 2133 MHz

In other words; Turbo is working but your turbo settings in the BIOS are set to 16/16/23/24 aka 2133/2133/3066/2400 MHz and that is, apparently, not what you want. Please change the first two and test again.

**Comment by benwwchen** [comment]
@Piker-Alpha Thanks so much for taking your precious time to help me. Also, Congratulations!

With the new version of the script, I still can't get turbo boost working. (14 is the highest. The 15 and 16 steps were achieved by Clover's auto-generated ssdt.)
Here is the ssdt.dsl generated by the v18.3 script:
[ssdt.dsl.v18.3.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/276263/ssdt.dsl.v18.3.zip)

I tried looking for turbo settings but can only find the Enable/Disable option. I think 16/16/23/24 aka 2133/2133/3066/3200 comes with the CPU. And with that default same setting, in Windows it can reach various combinations of clock speed as shown below:
![clockspeedwin](https://cloud.githubusercontent.com/assets/4254675/15453647/f2e461dc-2050-11e6-81e5-9a910797aa40.png)

Also, Geekbench single-core score in Windows is about 1.5 times of that in OS X while multi-core scores are the same in Windows and OS X. So it may be the latter two, 23(2 cores)/24(1 core), that has problems in OS X.

I'll wait until you have more free time.

**Comment by Piker-Alpha** [comment]
Thanks. Now. The turbo frequencies MSR can be set to anything that you like, and perhaps that is what Windows does. Try a run with:

`./ssdtPRGen.sh -mode custom -p L3426 -x 1`

After that you add the -xcpm boot argument.

Note: I used the -mode custom arguments only because I do not have this CPU. You on the other hand should not need it.

**Comment by benwwchen** [comment]
I'm sorry that I've been leaving home for school so I can't test it until weekends. I'll reply as soon as I try it. Thanks again.

**Comment by benwwchen** [comment]
Just tried adding `-x 1` and boot with -xcpm but it stucks at Low Frequency Mode (1.2GHz).
Here is the ssdt.dsl generated:
[ssdt.xcpm.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/286502/ssdt.xcpm.zip)

Here is the output of the script:

```
./ssdtPRGen.sh -p L3426 -x 1

ssdtPRGen.sh v0.9  Copyright (c) 2011-2012 by † RevoGirl
             v6.6  Copyright (c) 2013 by † Jeroen
             v18.3 Copyright (c) 2013-2016 by Pike R. Alpha
-----------------------------------------------------------
Bugs > https://github.com/Piker-Alpha/ssdtPRGen.sh/issues <

System information: Mac OS X 10.11.5 (15F34)
Brandstring: "Intel(R) Xeon(R) CPU L3426 @ 1.87GHz"

Override value: (-p) processor model, now using: L3426!
Override value: (-x) XCPM mode, now set to: 1!



Scope (_PR_) {592 bytes} with ACPI Processor declarations found in the DSDT (ACPI 1.0 compliant)
Generating ssdt.dsl for a 'MacPro5,1' with board-id [Mac-F221BEC8]
Intel L3426 processor [0x106E5] setup [0x0501]
With a maximum TDP of 45 Watt, as specified by Intel
Number logical CPU's: 8 (Core Frequency: 1867 MHz)
Number of Turbo States: 10 (2000-3200 MHz)
Number of P-States: 16 (1200-3200 MHz)
Injected C-States for P001 (C1,C3,C6)
Warning: 'cpu-type' may be set improperly (0x0501 instead of 0x01)
Warning: 'board-id' [Mac-F221BEC8] is not supported by Unknown power management
Warning: 'system-type' may be set improperly (1 instead of 0)

Compiling: ssdt_pr.dsl
Intel ACPI Component Architecture
ASL Optimizing Compiler version 20140926-64 [Nov  6 2014]
Copyright (c) 2000 - 2014 Intel Corporation

ASL Input:     /Users/benwwchen/Library/ssdtPRGen/ssdt.dsl - 225 lines, 6151 bytes, 57 keywords
AML Output:    /Users/benwwchen/Library/ssdtPRGen/ssdt.aml - 1558 bytes, 21 named objects, 36 executable opcodes

Compilation complete. 0 Errors, 0 Warnings, 0 Remarks, 0 Optimizations
```

Here is the output of AppleIntelInfo.kext:

```
AppleIntelInfo.kext v1.2 Copyright © 2012-2015 Pike R. Alpha. All rights reserved

Settings:
------------------------------------
logMSRs............................: 0
logIGPU............................: 0
logIntelRegs.......................: 1
logCStates.........................: 1
logIPGStyle........................: 1
InitialTSC.........................: 0x420195401a
MWAIT C-States.....................: 4384

CPU Ratio Info:
------------------------------------
CPU Low Frequency Mode.............: 900 MHz
CPU Maximum non-Turbo Frequency....: 1400 MHz
CPU Maximum Turbo Frequency........: 2400 MHz
CPU P-States [ ]
CPU C3-Cores [ 0 1 2 3 4 5 7 ]
CPU C6-Cores [ 0 1 2 3 ]
CPU P-States [ 9 ]
CPU C3-Cores [ 0 1 2 3 4 5 6 7 ]
CPU C6-Cores [ 0 1 2 3 4 5 ]
CPU C6-Cores [ 0 1 2 3 4 5 6 7 ]
```

Would it be the reason that OS X will always use 4 cores [according to Slice here](http://www.insanelymac.com/forum/topic/284656-clover-general-discussion/?p=2239825) so the turboboost can only boost 4 cores together to 2.13GHz?

**Comment by Piker-Alpha** [comment]
The initial issue with the script – being that it failed to produce a working SSDT – is fixed. That is good. 

You also said that the Clover generated SSDT can turbo to 2.13 GHz, but the attached copy of: "System SSDT.zip" in the first post is not the one I need to compare the two (the one from Clover and mine). Please attach the correct one (with the P-States in it).

I don't understand why "Slice" thinks that OS X uses four cores all the time, because that is certainly not the case. However. He may have meant to say something else. Like I said. The turbo MSR is set so that three and four active cores use a multiplier of 16, instead of a higher one. Like MS Windows seem to do.

And yeah. I said to use the -xcpm boot argument, but that is not supported with your processor. Only later processors, so please stop using it. Or you end up with the notorious error, that it is stuck at the lowest possible frequency.

**Comment by benwwchen** [comment]
@Piker-Alpha 
The "System SSDT.zip" in the first post is the only one SSDT I can find from MaciASL when 2.13 GHz boosting correctly.

If I select "Generate PStates" and "Generate CStates" in Clover, there will be 2 more ssdt I can get from MaciASL, but I can only get 1.86 GHz when selecting the two options in Clover, just like I'm using the SSDT generated by your script.

That is to say, it seems that when the CPU can turbo to 2.13 GHz, the SSDT contains no P-States. When the SSDT contains P-States, the CPU cannot turboboost at all (with only 1.86 GHz highest).

**Comment by Piker-Alpha** [comment]
I see. OK. Please open IORegistryExplorer and check if ACPI_SMC_PlatformPlugin or X86PlatformPlugin is loaded (see CPU0/P001). Also check the number of CPUPStates and the highest/lowest Frequency. Thanks!

**Comment by benwwchen** [comment]
ACPI_SMC_PlatformPlugin is loaded.
Are these numbers below what you need?

```
<4c 07 00 00 d8 d6 00 00 0a 00 00 00 0a 00 00 00 00 0f 00 00 00 0f 00 00>
<4b 07 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 0e 00 00 00 0e 00 00>
<c5 06 00 00 b8 88 00 00 0a 00 00 00 0a 00 00 00 00 0d 00 00 00 0d 00 00>
<40 06 00 00 78 69 00 00 0a 00 00 00 0a 00 00 00 00 0c 00 00 00 0c 00 00>
<bb 05 00 00 08 52 00 00 0a 00 00 00 0a 00 00 00 00 0b 00 00 00 0b 00 00>
<35 05 00 00 80 3e 00 00 0a 00 00 00 0a 00 00 00 00 0a 00 00 00 0a 00 00>
<b0 04 00 00 f8 2a 00 00 0a 00 00 00 0a 00 00 00 00 09 00 00 00 09 00 00>
```

(7 states from 1.20 GHz to ~~2.13 GHz~~ 1.868 GHz.) (Sorry I didn't check it carefully.)
![screenshot1](https://cloud.githubusercontent.com/assets/4254675/15628313/b659ad50-2530-11e6-9430-3d6b186796f3.png)
![screenshot2](https://cloud.githubusercontent.com/assets/4254675/15628314/b97ee5d6-2530-11e6-9781-e809fdae62bd.png)

**Comment by Piker-Alpha** [comment]
Is that what Clover generated? I mean this is what you have:

1868 MHz (0x74c)
1864 MHz (0x74b)
1733 MHz (0x6C5)
1600 MHz (0x0640)
1467 MHz (0x5BB)
1333 MHz (0x535)
1200 MHz (0x04b0)

No Turbo P-States whatsoever. The generated SSDT (with ssdtPRGen.sh) has all turbo P-States defined.

Edit: Fixed two typos (0x335 should have been 0x535 and 1733MHz was missing)

**Comment by benwwchen** [comment]
Yes. When I check those numbers in IORegistryExplorer, I'm using the SSDT without P-States (the "System SSDT" one) and the CPU can turbo to 2.13GHz at that time. It's so weird that the states in IORegistryExplorer don't match what I actually get.

When I use the generated SSDT (with ssdtPRGen.sh), I can see all states below in IORegistryExplorer but actually I can only get 1.86 GHz.

```
<80 0c 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 18 00 00 00 18 00 00>
<fa 0b 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 17 00 00 00 17 00 00>
<75 0b 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 16 00 00 00 16 00 00>
<f0 0a 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 15 00 00 00 15 00 00>
<6a 0a 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 14 00 00 00 14 00 00>
<e5 09 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 13 00 00 00 13 00 00>
<60 09 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 12 00 00 00 12 00 00>
<db 08 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 11 00 00 00 11 00 00>
<55 08 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 10 00 00 00 10 00 00>
<d0 07 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 0f 00 00 00 0f 00 00>
<4b 07 00 00 c8 af 00 00 0a 00 00 00 0a 00 00 00 00 0e 00 00 00 0e 00 00>
<c5 06 00 00 25 80 00 00 0a 00 00 00 0a 00 00 00 00 0d 00 00 00 0d 00 00>
<40 06 00 00 b9 65 00 00 0a 00 00 00 0a 00 00 00 00 0c 00 00 00 0c 00 00>
<bb 05 00 00 36 4f 00 00 0a 00 00 00 0a 00 00 00 00 0b 00 00 00 0b 00 00>
<36 05 00 00 4b 3c 00 00 0a 00 00 00 0a 00 00 00 00 0a 00 00 00 0a 00 00>
<b0 04 00 00 a7 2c 00 00 0a 00 00 00 0a 00 00 00 00 09 00 00 00 09 00 00>
```

**Comment by Piker-Alpha** [comment]
Ok so with seven p-states it sort of works, then replace this section in your generated SSDT and try again:

```
        Name (APLF, Zero)
        Name (APSN, One)
        Name (APSS, Package (0x07)
        {
            /* Workaround for the Ivy Bridge PM 'bug' */
            Package (0x06) { 0x074B, 0x00AFC8, 0x0A, 0x0A, 0x1300, 0x1300 },
            /* High Frequency Modes (non-turbo) */
            Package (0x06) { 0x074A, 0x00AFC8, 0x0A, 0x0A, 0x000E, 0x000E },
            Package (0x06) { 0x06C5, 0x008025, 0x0A, 0x0A, 0x000D, 0x000D },
            Package (0x06) { 0x0640, 0x0065B9, 0x0A, 0x0A, 0x000C, 0x000C },
            Package (0x06) { 0x05BB, 0x004F36, 0x0A, 0x0A, 0x000B, 0x000B },
            Package (0x06) { 0x0536, 0x003C4B, 0x0A, 0x0A, 0x000A, 0x000A },
            /* Low Frequency Mode */
            Package (0x06) { 0x04B0, 0x002CA7, 0x0A, 0x0A, 0x0009, 0x0009 }
        })
```

Please ignore the Ivy Bridge comment, this was generated with a modified copy of ssdtPRGen.sh

**Comment by benwwchen** [comment]
It still can't turbo but stuck at 1.86 GHz. (Sorry for being only able to test it at weekends.)
Here is the SSDT in which I've replaced the section you referred to:
[ssdt.replaced.zip](https://github.com/Piker-Alpha/ssdtPRGen.sh/files/297023/ssdt.replaced.zip)

**Comment by Piker-Alpha** [comment]
Me too. I can only work at weekends on hack related stuff, so that is no problem for me. Try this:

```
        Name (APSS, Package (0x07)
        {
            Package (0x06) { 0x074C, 0x00D6D8, 0x0A, 0x0A, 0x000F, 0x000F },
            Package (0x06) { 0x074B, 0x00AFC8, 0x0A, 0x0A, 0x000E, 0x000E },
            Package (0x06) { 0x06C5, 0x0088B8, 0x0A, 0x0A, 0x000D, 0x000D },
            Package (0x06) { 0x0640, 0x006978, 0x0A, 0x0A, 0x000C, 0x000C },
            Package (0x06) { 0x05BB, 0x005208, 0x0A, 0x0A, 0x000B, 0x000B },
            Package (0x06) { 0x0536, 0x003EB0, 0x0A, 0x0A, 0x000A, 0x000A },
            Package (0x06) { 0x04B0, 0x002AF8, 0x0A, 0x0A, 0x0009, 0x0009 }
        })
```

This is the same table that works with clover, but if that still isn't enough, then also rename all instances of APSS to _PSS and ACST to _CST and try again.

**Comment by benwwchen** [comment]
By replacing the table you provided above which works with clover, it can turboboost perfectly now!

```
AppleIntelInfo.kext v1.2 Copyright © 2012-2015 Pike R. Alpha. All rights reserved

Settings:
------------------------------------
logMSRs............................: 0
logIGPU............................: 0
logIntelRegs.......................: 1
logCStates.........................: 1
logIPGStyle........................: 1
InitialTSC.........................: 0x8663610ff4
MWAIT C-States.....................: 4384

CPU Ratio Info:
------------------------------------
CPU Low Frequency Mode.............: 900 MHz
CPU Maximum non-Turbo Frequency....: 1400 MHz
CPU Maximum Turbo Frequency........: 2400 MHz
CPU P-States [ 12 ]
CPU C3-Cores [ 0 2 4 6 7 ]
CPU C6-Cores [ 0 1 2 3 4 5 6 7 ]
CPU C3-Cores [ 0 1 2 3 4 5 6 7 ]
CPU P-States [ 12 15 ]
CPU P-States [ 12 15 18 ]
CPU P-States [ 12 15 18 21 ]
CPU P-States [ 12 15 18 21 23 ]
CPU P-States [ 12 15 16 18 21 23 ]
CPU P-States [ 12 13 15 16 18 21 23 ]
CPU P-States [ 12 13 14 15 16 18 21 23 ]
CPU P-States [ 11 12 13 14 15 16 18 21 23 ]
CPU P-States [ 11 12 13 14 15 16 17 18 21 23 ]
CPU P-States [ 11 12 13 14 15 16 17 18 20 21 23 ]
CPU P-States [ 10 11 12 13 14 15 16 17 18 20 21 23 ]
CPU P-States [ 10 11 12 13 14 15 16 17 18 20 21 22 23 ]
CPU P-States [ 9 10 11 12 13 14 15 16 17 18 20 21 22 23 ]
CPU P-States [ 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 ]
CPU P-States [ 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 ]
```

The CPU can reach all states and Geekbench scores are nearly the same as those in Windows now.
(I didn't rename APSS or ACST because renaming ACST to _CST causes a compilation error and if only renaming APSS to _PSS, it would make the CPU only turbo to 2.13 GHz again.)

But I'm curious about why the table with only 7 states in it could work?

Thanks so much for your patience and great help!

**Comment by Piker-Alpha** [comment]
Great, but oops. Looks like my 0x1300, 0x1300 versus 0x000F, 0x000F was the problem in my previously provided table. Sorry about that. Can you please verify that for me?

I would also replace the frequencies  (in the first column) and power (in the second column) and see if that works, because the power is now set to 55 Watt instead of 45 Watt. 

Now my next task is to make ssdtPRGen.sh work with your CPU ;) Oh and back in the day it used to work like this, with less data for turbo frequencies.

**Comment by benwwchen** [comment]
Changing "0x1300, 0x1300" to "0x000F, 0x000F" also makes the previous table work like a charm. All states can be reached with the previous table.

What should frequencies and power be changed to? Or I misunderstood something?

**Comment by Piker-Alpha** [comment]
Nope. Nothing. It is like I said. I forgot to fix the 0x1300 in my previous table. Closing issue.
